### ISSUE TITLE
Chezmoi is not purging unmanaged files with externals

### ISSUE THREAD

**Issue by joshuaspence** [twpayne/chezmoi#3703]
## Describe the bug

I use externals to manage a bunch of files in `~/.local/bin` and I want chezmoi to remove any files in this directory that don't exist in the source state, so I created an empty file called `dot_local/exact_bin/.keep`. But chezmoi does not remove unmanaged files.

## To reproduce

```
# .chezmoiexternal.yaml
---
.local/bin/chezmoi:
  type: 'file'
  executable: true
  url: '{{ (gitHubLatestRelease "twpayne/chezmoi").HTMLURL | replace "/releases/tag/" "/releases/download/" }}/{{ printf "chezmoi-%s-%s" .chezmoi.os .chezmoi.arch }}'
```
```
# dot_local/exact_bin/.keep
```

With unmanaged files in `~/.local/bin`, run `chezmoi diff`.

## Expected behavior

`chezmoi diff` would show that the unmanaged files would be removed.

## Output of command with the `--verbose` flag

```console
$ chezmoi --verbose diff
# Literally empty
```

## Output of `chezmoi doctor`

<details>

```console
$ chezmoi doctor
RESULT    CHECK                       MESSAGE
ok        version                     v2.47.4, commit d6ad485c86aa6501f5a7754ccba2a2b03d46c1a2, built at 2024-04-12T22:51:34Z, built by goreleaser
ok        latest-version              v2.47.4
ok        os-arch                     linux/amd64 (Ubuntu 22.04.4 LTS (Jammy Jellyfish))
ok        uname                       Linux jspence-Precision-5680 6.5.0-17-generic #17~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Jan 16 14:32:32 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
ok        go-version                  go1.22.2 (gc)
ok        executable                  ~/.local/bin/chezmoi
ok        upgrade-method              replace-executable
ok        config-file                 ~/.config/chezmoi/chezmoi.yaml, last modified 2024-02-21T16:34:32+11:00
warning   source-dir                  ~/.local/share/chezmoi is a git working tree (dirty)
ok        suspicious-entries          no suspicious entries
warning   working-tree                ~/.local/share/chezmoi is a git working tree (dirty)
ok        dest-dir                    ~ is a directory
ok        umask                       002
ok        cd-command                  found /bin/bash
ok        cd-args                     /bin/bash
info      diff-command                not set
ok        edit-command                found /usr/bin/vim
ok        edit-args                   /usr/bin/vim
ok        git-command                 found /usr/bin/git, version 2.34.1
ok        merge-command               found /usr/bin/vimdiff
ok        shell-command               found /bin/bash
ok        shell-args                  /bin/bash
ok        age-command                 found ~/go/bin/age, version 1.1.1
ok        gpg-command                 found /usr/bin/gpg, version 2.2.27
info      pinentry-command            not set
ok        1password-command           found /usr/bin/op, version 2.26.1
info      bitwarden-command           bw not found in $PATH
info      bitwarden-secrets-command   bws not found in $PATH
info      dashlane-command            dcli not found in $PATH
info      doppler-command             doppler not found in $PATH
info      gopass-command              gopass not found in $PATH
info      keepassxc-command           keepassxc-cli not found in $PATH
info      keepassxc-db                not set
info      keeper-command              keeper not found in $PATH
info      lastpass-command            lpass not found in $PATH
info      pass-command                pass not found in $PATH
info      passhole-command            ph not found in $PATH
info      rbw-command                 rbw not found in $PATH
info      vault-command               vault not found in $PATH
info      vlt-command                 vlt not found in $PATH
info      secret-command              not set
```

</details>

## Additional context

This seems to have been broken by #3428. Using chezmoi v2.42.3 I get the expected behaviour.

**Comment by twpayne** [comment]
> But chezmoi does not remove unmanaged files.

chezmoi is behaving as expected here. The file `.local/bin/chezmoi` is managed by chezmoi. If the URL changes (either the URL is changed to the data returned from the URL changes) then chezmoi replace the contents of `.local/bin/chezmoi`, i.e. there are no files to add or remove in the `exact_` `.local/bin` directory.

**Comment by twpayne** [comment]
Thanks for reporting this. It was a very confusing bug. It should be fixed with #3705.
